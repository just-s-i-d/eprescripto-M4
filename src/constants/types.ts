export type InfoCardDetailsType =
  | {
      totalCount: number;
      title: string;
    }[]
  | undefined;

export type ErrorType = {
  message: string;
  type: string;
};
export type AppointmentDataType = {
  pId: string;
  email: string;
  lastVisited: string;
  avatar: string;
  gender: string;
  key: number;
  pName: string;
  timeSlot: string;
  contact: string;
  status: string;
  profilePic: string;
  referrer: string;
  date: string;
};

export type AppointmentsDataType = AppointmentDataType[] | undefined;

export type PrescriptionDataType = {
  key: React.Key;
  prescriptionId: string;
  name: string;
  duration: number;
  date: string;
  status: "Active" | "Completed" | "On Hold";
  notes: string;
};

export type PrescrtionsDataType = PrescriptionDataType[] | undefined;

export type ChartDataType = {
  labels: string[];
  data: number[];
  totalCount: number;
};

export type LineChartDataType = {
  labels: string[];
  data: number[];
};

export type LineChartApiResDataType = {
  seven: number[];
  ten: number[];
  thirty: number[];
};

export type ChartApiDataType = {
  pieChartData: ChartDataType;
  barChartData: ChartDataType;
  lineChartData: LineChartApiResDataType;
};

export type DoctorAppointmentsDataType = AppointmentDataType[];

export type ReviewDataType = {
  pId: string;
  pName: string;
  email: string;
  rating: 1 | 2 | 3 | 4 | 5;
  comment: string;
};
export type ReviewsDataType = ReviewDataType[];
