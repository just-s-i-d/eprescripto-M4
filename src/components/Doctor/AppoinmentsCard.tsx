import { useState } from "react";
import { Card, Skeleton, Avatar, Tooltip, Image } from "antd";
import { CloseCircleTwoTone, PhoneTwoTone } from "@ant-design/icons";

import noDataPng from "@assets/no-data.png";
import CardTitle from "@components/ui/CardTitle";
import PopModal from "@components/ui/PopModal";
import profilePic from "@assets/profilepic.png";
import {
  ApiResponseData,
  ApiResponseDataType,
  AppointmentDataType,
} from "@constants/types";
import { cancelAppointment } from "@utils/Doctor";
import ErrorBoundary from "@components/ErrorBoundary";
import useStatesHook from "src/hooks/useStatesHook";
import { showToast } from "@utils/common";

type AppointmentCardPropsType = {
  appointments?: ApiResponseData<AppointmentDataType>;
  loading: boolean;
  appointmentsState: ReturnType<
    typeof useStatesHook<ApiResponseData<AppointmentDataType>>
  >;
  getDataForAppointments: () => void;
};

const AppoinmentsCard = ({
  appointments,
  loading,
  appointmentsState,
  getDataForAppointments,
}: AppointmentCardPropsType) => {
  const [open, setOpen] = useState<boolean>(false);
  const [selectedPatient, setSelectedPatient] =
    useState<ApiResponseDataType<AppointmentDataType>>();
  const handlerCancleAppointment = (
    patient: ApiResponseDataType<AppointmentDataType>,
  ) => {
    setOpen(true);
    setSelectedPatient(patient);
  };
  const confirmHandler = async () => {
    if (selectedPatient) {
      cancelAppointment(selectedPatient)
        .then((res) => {
          showToast(res.message, res.type);
          getDataForAppointments();
        })
        .catch((error) => {
          showToast(error.message, error.type);
        });
    }
  };
  return (
    <>
      <Card
        className={`w-[32%] max-xxl:h-[59vh] max-h-[59vh] max-xl:max-h-[60vh] max-xxl:w-[48%] max-xl:w-[48%] max-md:w-9/12 max-sm:w-full `}
        bordered={false}
      >
        <ErrorBoundary
          refreshComponent={() => appointmentsState.setRefresh((prev) => !prev)}
          error={appointmentsState.error}
        >
          <CardTitle>Today's Appointments</CardTitle>
          {!appointmentsState.loading && !appointmentsState.data?.length ? (
            <div className="flex flex-col items-center justify-center">
              <Image src={noDataPng} preview={false} />
              <h2 className="text-lg mt-12 text-center">No Appointments</h2>
            </div>
          ) : (
            <div className="mx-auto px-4 h-[44vh] overflow-y-auto">
              <ul className="mt-1 divide-y">
                {appointments?.map((item, idx) => (
                  <li
                    key={idx}
                    className="py-4 mb-2 px-2 bg-[#F0F3F4] rounded-lg"
                  >
                    <div className="flex items-start justify-between">
                      <div className="flex gap-3">
                        {loading ? (
                          <Skeleton.Avatar active size={48} />
                        ) : (
                          <Avatar
                            size={48}
                            src={item.attributes?.profilePic || profilePic}
                          />
                        )}
                        <Skeleton
                          loading={loading}
                          active
                          avatar={false}
                          title={false}
                          paragraph={{ rows: 2, width: [100, 100] }}
                        >
                          <div>
                            <span className="block text-base text-gray-700">
                              {item.attributes?.pName}
                            </span>
                            <span className="block text-[12px] text-gray-600">
                              {item.attributes?.timeSlot}
                            </span>
                          </div>
                        </Skeleton>
                      </div>
                      {loading ? (
                        <Skeleton.Button active />
                      ) : (
                        <span
                          className={`py-1 px-2 rounded-lg ${item.attributes?.status.toLowerCase()}`}
                        >
                          {item.attributes?.status}
                        </span>
                      )}
                    </div>
                    <div className="flex justify-between">
                      <span className="block text-sm text-gray-700 font-bold">
                        <span className="mr-1 py-1 text-gray-500 font-semibold">
                          Referrer
                        </span>
                        {item.attributes.referrer || "NA"}
                      </span>
                      <div className="flex justify-end gap-4 text-lg">
                        <Tooltip title={item.attributes?.contact}>
                          <PhoneTwoTone
                            className="rotate-90 cursor-pointer"
                            href={`tel:${item.attributes?.contact}`}
                          />
                        </Tooltip>

                        {item.attributes.status === "Pending" && (
                          <CloseCircleTwoTone
                            className="cursor-pointer"
                            onClick={() => handlerCancleAppointment(item)}
                          />
                        )}
                      </div>
                      <PopModal
                        setOpen={setOpen}
                        open={open}
                        okButtonText="Sure"
                        title="Cancel Appointment"
                        footer={true}
                        confirmHandler={confirmHandler}
                      >
                        Are you sure,you want to cancel the appoinment for{" "}
                        {selectedPatient?.attributes.pName}?
                      </PopModal>
                    </div>
                  </li>
                ))}
              </ul>
            </div>
          )}
        </ErrorBoundary>
      </Card>
    </>
  );
};

export default AppoinmentsCard;
