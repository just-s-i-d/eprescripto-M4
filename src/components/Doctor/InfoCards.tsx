import { useEffect, useState } from "react";
import { Row, Col, Card, Image, Typography, Skeleton } from "antd";

import ErrorBoundary from "@components/ErrorBoundary";
import CardTitle from "@components/ui/CardTitle";
import { ApiResponseData, InfoCardDetailsType } from "@constants/types";
import { CARD_PROPERTIES } from "@constants/constants";
import useStatesHook from "../../hooks/useStatesHook";

const { Text } = Typography;

type InfoCardsProps = {
  cardDetails?: ApiResponseData<InfoCardDetailsType>;
  infoCards: ReturnType<
    typeof useStatesHook<ApiResponseData<InfoCardDetailsType>>
  >;
};

const getBgColor = (index: number) => {
  switch (index + 1) {
    case 1:
      return "bg-infoCardColors-color1";
    case 2:
      return "bg-infoCardColors-color2";
    case 3:
      return "bg-infoCardColors-color3";
    case 4:
      return "bg-infoCardColors-color4";
  }
};
const InfoCards = ({ cardDetails, infoCards }: InfoCardsProps) => {
  const [loading, setLoading] = useState(true);
  useEffect(() => {
    cardDetails && setTimeout(() => setLoading(false), 500);
  }, [cardDetails]);
  return (
    <Card
      className="info-cards-body w-full h-[22vh] max-md:h-full flex justify-center"
      bordered={false}
    >
      <ErrorBoundary
        error={infoCards.error}
        refreshComponent={() => infoCards.setRefresh((prev) => !prev)}
      >
        <Row
          gutter={30}
          className="w-full max-md:flex max-md:flex-wrap gap-y-6 justify-center max-lg:flex-nowrap"
        >
          {cardDetails?.map((element, index) => {
            return (
              <ErrorBoundary key={element.attributes.title}>
                <Col className="basis-3/12 max-md:basis-6/12 max-sm:basis-9/12">
                  <Card
                    className={`info-card h-[17vh] py-8 px-4 max-sm:py-5 max-w-[390px] max-xl:h-[16vh] max-xl:min-w-[135px] max-sm:h-[150px] ${getBgColor(
                      index,
                    )} text-white`}
                    bordered={false}
                  >
                    <div>
                      <div className={`flex items-center gap-4 mb-2 mt-2`}>
                        {loading ? (
                          <Skeleton.Avatar active shape="square" size={50} />
                        ) : (
                          <Image
                            src={CARD_PROPERTIES[index]}
                            alt="Icon image"
                            className="w-[50px] max-xxl:w-[35px] max-md:w-[40px]"
                            preview={false}
                          />
                        )}
                        {loading ? (
                          <Skeleton.Button />
                        ) : (
                          <Text className="text-4xl max-xl:text-3xl text-white max-md:text-4xl">
                            {element.attributes.totalCount}
                          </Text>
                        )}
                      </div>
                      {loading ? (
                        <Skeleton.Input active />
                      ) : (
                        <CardTitle className="max-xxl:text-lg max-xl:text-base max-md:text-lg">
                          {element.attributes.title}
                        </CardTitle>
                      )}
                    </div>
                  </Card>
                </Col>
              </ErrorBoundary>
            );
          })}
        </Row>
      </ErrorBoundary>
    </Card>
  );
};
export default InfoCards;
