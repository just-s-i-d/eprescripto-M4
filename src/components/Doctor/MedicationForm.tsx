import { Input, Select, Form } from "antd";
import { CloseOutlined } from "@ant-design/icons";
import { MedicationType } from "@constants/types";

const { Item } = Form;
type MedicationFormPropsType = {
  fields: MedicationType;
  removeable?: boolean;
  removeMedicationInput: (id: number) => void;
};
const MedicationForm = ({
  fields,
  removeMedicationInput,
  removeable,
}: MedicationFormPropsType) => {
  return (
    <>
      <div className="mt-4 mb-2 flex justify-between">
        <h2 className="text-xl">Medication</h2>
        {removeable && (
          <CloseOutlined
            onClick={() => removeMedicationInput(fields.id)}
            data-testid="remove-field"
            className="text-[16px]"
          />
        )}
      </div>

      <div className="w-full grid sm:grid-cols-2 md:grid-cols-4 gap-x-5 gap-y-3">
        <Item
          name={fields.medicine}
          label="Medicine Name"
          className="w-full"
          required
          rules={[
            {
              required: true,
              message: "Field cannot be empty",
            },
            {
              pattern: /^[a-zA-Z0-9.-]+$/,
              message: "Special characters not allowed",
            },
          ]}
        >
          <Input />
        </Item>
        <Item
          name={fields.dosage}
          label="Dosage"
          className="w-full"
          required
          rules={[
            {
              required: true,
              message: "Field cannot be empty",
            },
            {
              pattern: /^[0-9/mg]/,
              message: "Enter a valid dosage",
            },
          ]}
        >
          <Input />
        </Item>
        <Item
          name={fields.timesPerDay}
          label="Frequency"
          className="w-full"
          required
          rules={[
            {
              required: true,
              message: "Field cannot be empty",
            },
          ]}
        >
          <Select
            options={[
              {
                value: 1,
                label: "1/Day",
              },
              {
                value: 2,
                label: "2/Day",
              },
              {
                value: 3,
                label: "3/Day",
              },
            ]}
          />
        </Item>
        <div className="flex gap-6">
          <Item
            name={fields.instruction}
            className="w-full"
            label="Instruction"
            required
            rules={[
              {
                required: true,
                message: "Field cannot be empty",
              },
            ]}
          >
            <Select
              options={[
                {
                  value: "beforeFood",
                  label: "Before eating food",
                },
                {
                  value: "afterFood",
                  label: "After eating Food",
                },
              ]}
            />
          </Item>
        </div>
      </div>
    </>
  );
};
export default MedicationForm;
