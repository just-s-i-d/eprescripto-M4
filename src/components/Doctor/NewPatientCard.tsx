import ErrorBoundary from "@components/ErrorBoundary";
import { Form, FormInstance, Input, Select } from "antd";
import FormItem from "antd/es/form/FormItem";
import TextArea from "antd/es/input/TextArea";

type NewPatientCardPropType = {
  newPatientForm: FormInstance;
};
const NewPatientCard = ({ newPatientForm }: NewPatientCardPropType) => {
  return (
    <ErrorBoundary>
      <Form form={newPatientForm} layout="vertical" className="mt-6 px-4">
        <FormItem
          name="pName"
          label="Patient Name"
          required
          rules={[
            {
              required: true,
              message: "Field cannot be empty",
            },
            {
              pattern: /^[A-Za-z\s]+$/,
              message: "Special characters not allowed",
            },
          ]}
        >
          <Input />
        </FormItem>
        <FormItem
          name="email"
          label="Email"
          required
          rules={[
            {
              required: true,
              message: "Field cannot be empty",
            },
            {
              pattern: /^[\w.-]+@[\w.-]+\.\w+$/,
              message: "Enter a valid mail",
            },
          ]}
        >
          <Input />
        </FormItem>
        <div className="flex gap-6">
          <FormItem
            name="dob"
            label="Date of Birth"
            required
            className="w-full"
            rules={[
              {
                required: true,
                message: "Please enter your age!",
              },
            ]}
          >
            <Input type="date" />
          </FormItem>
          <FormItem
            name="gender"
            label="Gender"
            required
            className="w-full"
            rules={[
              {
                required: true,
                message: "Please select your gender!",
              },
            ]}
          >
            <Select
              placeholder="Select your gender"
              options={[
                { value: "male", label: "Male" },
                { value: "female", label: "Female" },
                { value: "other", label: "Other" },
              ]}
            />
          </FormItem>
        </div>
        <FormItem
          name="contactNo"
          label="Mobile"
          required
          rules={[
            {
              required: true,
              message: "Field cannot be empty",
            },
            {
              pattern: /^[6-9]\d{9}$/,
              message: "Enter valid no.",
            },
          ]}
        >
          <Input />
        </FormItem>
        <FormItem name="allergies" label="Mention allergies , if any">
          <TextArea placeholder="Mention your allergies here" />
        </FormItem>
      </Form>
    </ErrorBoundary>
  );
};
export default NewPatientCard;
