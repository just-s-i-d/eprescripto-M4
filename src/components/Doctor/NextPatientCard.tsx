import { useEffect, useState } from "react";

import { Card, Image, Space, Skeleton, Button } from "antd";

import CardTitle from "@components/ui/CardTitle";
import PopModal from "@components/ui/PopModal";
import TableCard from "@components/ui/TableCard";
import ErrorBoundary from "@components/ErrorBoundary";
import noDataPng from "@assets/no-data.png";
import profilePic from "@assets/profilepic.png";
import {
  ApiResponseData,
  AppointmentApiResponseDataType,
  AppointmentsApiResonseType,
  GenericObjectType,
  PrescriptionDataType,
} from "@constants/types";
import {
  cancelAppointment,
  formatDateReadable,
  getData,
  prescriptionsDataEndPoint,
} from "@utils/Doctor";
import useStatesHook from "../../hooks/useStatesHook";
import { columns } from "@constants/constants";
import { showToast } from "@utils/common";
import { ColumnGroupType } from "antd/es/table";

type NextPatientCardProps = {
  appointment?: AppointmentApiResponseDataType;
  error: boolean;
  appointments: ReturnType<typeof useStatesHook<AppointmentsApiResonseType>>;
  getDataForAppointments: () => void;
};

const NextPatientCard = ({
  appointment,
  appointments,
  getDataForAppointments,
}: NextPatientCardProps) => {
  const [open, setOpen] = useState(false);
  const [visibililty, setVisibility] = useState(false);
  const prescriptions = useStatesHook();
  const setPrescriptionsData = () => {
    getData<ApiResponseData<PrescriptionDataType>>(prescriptionsDataEndPoint)
      .then((res) => {
        const data = res.map((element) => element.attributes);
        prescriptions.setData(data);
        prescriptions.setLoading(false);
        prescriptions.setError(false);
      })
      .catch(() => {
        prescriptions.setError(true);
      });
  };
  const handlePrescriptionClick = async () => {
    setVisibility(true);
    prescriptions.setLoading(true);
    setPrescriptionsData();
  };
  const confirmHandler = async () => {
    if (appointment) {
      cancelAppointment(appointment)
        .then((res) => {
          showToast(res.message, res.type);
          getDataForAppointments();
        })
        .catch((error) => {
          showToast(error.message, error.type);
        });
    }
  };
  useEffect(() => {
    setPrescriptionsData();
  }, [prescriptions.refresh]); //eslint-disable-line
  return (
    <Card
      className="xs:w-full sm:w-9/12 md:w-[48%] xxl:w-[32%]"
      bordered={false}
    >
      <ErrorBoundary
        error={appointments.error}
        refreshComponent={() => appointments.setRefresh((prev) => !prev)}
      >
        <CardTitle>Next Patient</CardTitle>
        {!appointments.loading && !appointments.data?.length ? (
          <div className="flex flex-col items-center justify-center">
            <Image src={noDataPng} preview={false} />
            <h2 className="text-lg mt-12 text-center">No Appointment</h2>
          </div>
        ) : (
          <div className="flex flex-col items-center">
            {appointments.loading ? (
              <Skeleton.Avatar active size={90} />
            ) : (
              <Image
                src={appointment?.attributes.profilePic || profilePic}
                preview={false}
                width={100}
                className="rounded-[50%]"
              />
            )}
            <Space direction="vertical" align="center" className="mt-2">
              {appointments.loading ? (
                <Skeleton.Input active />
              ) : (
                <span className="xxl:text-[22px] max-xl:text-[18px] font-extrabold">
                  {appointment?.attributes.pName}
                </span>
              )}
              {appointments.loading ? (
                <Skeleton.Input active />
              ) : (
                <span className="text-base font-bold">
                  {appointment?.attributes.email}
                </span>
              )}
            </Space>
            <Space className="flex items-start text-lg mt-2">
              <Skeleton
                avatar={false}
                title={false}
                paragraph={{ rows: 6, width: [250, 250, 250, 250, 250, 250] }}
                active
                loading={appointments.loading}
              >
                <table className="text-base">
                  <tbody>
                    <tr>
                      <td className="px-2 py-1 text-gray-500 font-semibold">
                        Patient Id
                      </td>
                      <td className="px-2 py-1 text-lg">
                        {appointment?.attributes.pId}
                      </td>
                    </tr>
                    <tr>
                      <td className="px-2 py-1 text-gray-500 font-semibold">
                        Time Slot
                      </td>
                      <td className="px-2 py-1 text-lg">
                        {appointment?.attributes.timeSlot}
                      </td>
                    </tr>
                    <tr>
                      <td className="px-2 py-1 text-gray-500 font-semibold">
                        Last Visited
                      </td>
                      <td className="px-2 py-1 text-lg">
                        {appointment &&
                          formatDateReadable(
                            appointment.attributes.lastVisited,
                          )}
                      </td>
                    </tr>
                    <tr>
                      <td className="px-2 py-1 text-gray-500 font-semibold">
                        Phone
                      </td>
                      <td className="px-2 py-1 text-lg">
                        {appointment?.attributes.contact}
                      </td>
                    </tr>
                    <tr>
                      <td className="px-2 py-1 text-gray-500 font-semibold">
                        Gender
                      </td>
                      <td className="px-2 py-1 text-lg">
                        {appointment?.attributes.gender}
                      </td>
                    </tr>
                  </tbody>
                </table>
                <Space className="mt-5">
                  <Button onClick={() => setOpen(true)} danger>
                    Cancel Appointment
                  </Button>
                  <Button onClick={handlePrescriptionClick}>
                    Prescriptions
                  </Button>
                </Space>
              </Skeleton>
            </Space>
          </div>
        )}

        <PopModal
          footer={true}
          setOpen={setOpen}
          open={open}
          okButtonText="Sure"
          title="Cancel Appointment"
          confirmHandler={confirmHandler}
        >
          Are you sure,you want to cancel the appoinment for{" "}
          {appointment?.attributes.pName}?
        </PopModal>
        <PopModal
          className="width-1000"
          footer={false}
          title="All Prescriptions"
          open={visibililty}
          setOpen={setVisibility}
        >
          <ErrorBoundary>
            <TableCard
              tableData={prescriptions.data as GenericObjectType[]}
              columns={columns as ColumnGroupType<GenericObjectType>[]}
              error={prescriptions.error}
              setRefresh={prescriptions.setRefresh}
              loading={prescriptions.loading}
              setLoading={prescriptions.setLoading}
            />
          </ErrorBoundary>
        </PopModal>
      </ErrorBoundary>
    </Card>
  );
};
export default NextPatientCard;
