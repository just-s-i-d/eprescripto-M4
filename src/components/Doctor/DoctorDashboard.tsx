import { useEffect, useState } from "react";

import NextPatientCard from "./NextPatientCard";
import InfoCards from "./InfoCards";
import ErrorBoundary from "@components/ErrorBoundary";
import PieChart from "../Charts/PieChart";
import LineChart from "@components/Charts/LineChart";
import BarChart from "@components/Charts/BarChart";
import AppoinmentsCard from "./AppoinmentsCard";
import {
  AppointmentsDataType,
  ChartDataType,
  InfoCardDetailsType,
  LineChartApiResDataType,
} from "@constants/types";
import {
  barChartDataEndPoint,
  getAppointmentData,
  getChartData,
  getDataForLineGraph,
  getInfoCardsData,
  lineChartDataEndPoint,
  pieChartDataEndPoint,
} from "@utils/Doctor";

const DoctorDashboard: React.FC = () => {
  const [lineChartData, setLineChartData] = useState<LineChartApiResDataType>();
  const [infoCardError, setInfoCardError] = useState(false);
  const [appointmentCardError, setAppointmentCardError] = useState(false);
  const [appointments, setAppointments] = useState<AppointmentsDataType>();
  const [infoCardsData, setInfoCardsData] = useState<InfoCardDetailsType>();
  const [pieChartData, setPieChartData] = useState<ChartDataType>();
  const [barChartData, setBarChartData] = useState<ChartDataType>();

  useEffect(() => {
    getInfoCardsData()
      .then((res) => {
        setInfoCardsData(res);
        setInfoCardError(false);
      })
      .catch(() => setInfoCardError(true));
    getAppointmentData()
      .then((res) => setAppointments(res))
      .catch(() => setAppointmentCardError(true));
    getChartData(pieChartDataEndPoint).then((res) => {
      setPieChartData(res.pieChartData);
    });
    getChartData(lineChartDataEndPoint).then((res) => {
      setLineChartData(res.lineChartData);
    });
    getChartData(barChartDataEndPoint).then((res) => {
      setBarChartData(res.barChartData);
    });
  }, []);
  return (
    <>
      <ErrorBoundary>
        <div className="max-w-full flex">
          <div className="w-full flex flex-col gap-6">
            <ErrorBoundary error={infoCardError}>
              <InfoCards cardDetails={infoCardsData} error={infoCardError} />
            </ErrorBoundary>
            <div className="flex flex-wrap justify-between gap-y-6 max-tablet:justify-center">
              <ErrorBoundary>
                <NextPatientCard
                  appointment={appointments && appointments[0]}
                  error={appointmentCardError}
                />
              </ErrorBoundary>
              <ErrorBoundary>
                <PieChart
                  chartData={pieChartData}
                  title="Ratings Composition"
                  label="Reviews"
                />
              </ErrorBoundary>
              <ErrorBoundary>
                <AppoinmentsCard
                  appointments={appointments}
                  loading={appointmentCardError}
                />
              </ErrorBoundary>
              <ErrorBoundary>
                <LineChart
                  chartData={lineChartData}
                  title="Patients in Last 7 Days"
                  xAxesTitle="Dates"
                  yAxesTitle="Number of Patients"
                  getDataForLineGraph={getDataForLineGraph}
                />
              </ErrorBoundary>
              <ErrorBoundary>
                <BarChart
                  chartData={barChartData}
                  title="Age Groups Comparison"
                  label="Patients"
                  xAxesTitle="Different Age Groups (In Years)"
                  yAxesTitle="Number of Patients"
                />
              </ErrorBoundary>
            </div>
          </div>
        </div>
      </ErrorBoundary>
    </>
  );
};

export default DoctorDashboard;
