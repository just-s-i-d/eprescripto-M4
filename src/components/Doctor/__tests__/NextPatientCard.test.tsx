import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import NextPatientCard from "../NextPatientCard";
import useStatesHook from "@hooks/useStatesHook";
import { AppointmentsApiResonseType } from "@constants/types";
import { cancelAppointment, formatDateReadable } from "@utils/Doctor";
import { showToast } from "@utils/common";

jest.mock("@utils/Doctor", () => ({
  cancelAppointment: jest.fn(() =>
    Promise.resolve({ message: "Cancelled", type: "success" }),
  ),
  getData: jest.fn(() => Promise.resolve([])),
  prescriptionsDataEndPoint: "mocked/prescription-endpoint",
  formatDateReadable: jest.fn().mockReturnValue("28 Dec 2023"),
}));

jest.mock("@hooks/useStatesHook");
jest.mock("@utils/common", () => ({
  showToast: jest.fn(),
}));

const mockAppointment = {
  id: 24,
  attributes: {
    id: 24,
    pId: "A01",
    userId: "A01",
    lastVisited: "11/17/23",
    pName: "John doe",
    referrer: "Dr. Verma",
    gender: "Male",
    timeSlot: "10:00 am to 11:00 am",
    contact: "9876543210",
    status: "Pending",
    profilePic: "profile-pic-url",
    createdAt: "2023-12-27T11:32:46.415Z",
    updatedAt: "2023-12-27T11:32:47.335Z",
    publishedAt: "2023-12-27T11:32:47.332Z",
    date: "12/27/23",
    email: "abc@abc.com",
  },
};

describe("NextPatientCard", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  (useStatesHook as jest.Mock).mockReturnValue({
    data: [mockAppointment],
    setData: jest.fn(),
    loading: false,
    setLoading: jest.fn(),
    error: false,
    setError: jest.fn(),
    refresh: false,
    setRefresh: jest.fn(),
  });

  const mockAppointments = useStatesHook<AppointmentsApiResonseType>();
  it("Test to check if it renders correctly", () => {
    const { asFragment } = render(
      <NextPatientCard
        error={false}
        appointment={mockAppointment}
        appointments={mockAppointments}
        getDataForAppointments={jest.fn()}
      />,
    );
    expect(screen.getByText("Next Patient")).toBeInTheDocument();
    expect(screen.getByAltText("profile-pic")).toBeInTheDocument();
    expect(screen.getByText("Patient Id")).toBeInTheDocument();
    expect(screen.getByText("Gender")).toBeInTheDocument();
    expect(screen.getByText("Last Visited")).toBeInTheDocument();
    expect(screen.getByText("Phone")).toBeInTheDocument();
    expect(screen.getByText("Time Slot")).toBeInTheDocument();
    expect(
      screen.getByText(mockAppointment.attributes.pName),
    ).toBeInTheDocument();
    expect(
      screen.getByText(mockAppointment.attributes.email),
    ).toBeInTheDocument();
    expect(
      screen.getByText(mockAppointment.attributes.pId),
    ).toBeInTheDocument();
    expect(
      screen.getByText(mockAppointment.attributes.gender),
    ).toBeInTheDocument();
    expect(
      screen.getByText(
        formatDateReadable(mockAppointment.attributes.lastVisited),
      ),
    ).toBeInTheDocument();
    expect(
      screen.getByText(mockAppointment.attributes.contact),
    ).toBeInTheDocument();
    expect(
      screen.getByText(mockAppointment.attributes.timeSlot),
    ).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
  it("Error handle and try again btn test", async () => {
    render(
      <NextPatientCard
        error={false}
        appointment={mockAppointment}
        appointments={{ ...mockAppointments, error: true }}
        getDataForAppointments={jest.fn()}
      />,
    );
    await waitFor(() => {
      const tryAgainBtn = screen.getByText("Try again");
      expect(tryAgainBtn).toBeInTheDocument();
      fireEvent.click(tryAgainBtn);
      expect(mockAppointments.setRefresh).toHaveBeenCalled();
    });
  });
  it("Test to check if the appointment was cancelled successfully", async () => {
    render(
      <NextPatientCard
        error={false}
        appointment={mockAppointment}
        appointments={mockAppointments}
        getDataForAppointments={jest.fn()}
      />,
    );
    const cancelAppointmentBtn = screen.getByRole("button", {
      name: "Cancel Appointment",
    });
    fireEvent.click(cancelAppointmentBtn);
    const confirmBtn = screen.getByText("Sure");
    fireEvent.click(confirmBtn);
    await waitFor(() => {
      expect(cancelAppointment).toHaveBeenCalledWith(mockAppointment);
      expect(showToast).toHaveBeenCalledWith("Cancelled", "success");
    });
  });
  it("Test to check handling of error when appointment cancelletion fails", async () => {
    (cancelAppointment as jest.Mock).mockRejectedValue({
      message: "Error",
      type: "error",
    });
    render(
      <NextPatientCard
        error={false}
        appointment={mockAppointment}
        appointments={mockAppointments}
        getDataForAppointments={jest.fn()}
      />,
    );
    const cancelAppointmentBtn = screen.getByRole("button", {
      name: "Cancel Appointment",
    });
    fireEvent.click(cancelAppointmentBtn);
    const confirmBtn = screen.getByText("Sure");
    fireEvent.click(confirmBtn);
    await waitFor(() => {
      expect(cancelAppointment).toHaveBeenCalledWith(mockAppointment);
      expect(showToast).toHaveBeenCalledWith("Error", "error");
    });
  });
  it("Test to check if the precription button works properly or not", async () => {
    render(
      <NextPatientCard
        error={false}
        appointment={mockAppointment}
        appointments={mockAppointments}
        getDataForAppointments={jest.fn()}
      />,
    );
    await waitFor(() => {
      const prescriptionBtn = screen.getByRole("button", {
        name: "Prescriptions",
      });
      expect(prescriptionBtn).toBeInTheDocument();
      fireEvent.click(prescriptionBtn);
      expect(screen.getByText("All Prescriptions")).toBeInTheDocument();
    });
  });
});
