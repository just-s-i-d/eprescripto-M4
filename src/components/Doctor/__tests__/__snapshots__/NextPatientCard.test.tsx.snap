// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NextPatientCard Test to check if it renders correctly 1`] = `
<DocumentFragment>
  <div
    class="ant-card xs:w-full sm:w-9/12 md:w-[48%] xxl:w-[32%] css-dev-only-do-not-override-1adbn6x"
  >
    <div
      class="ant-card-body"
    >
      <h1
        class="text-xl font-bold mb-6 undefined"
      >
        Next Patient
      </h1>
      <div
        class="flex flex-col items-center"
      >
        <div
          class="ant-image css-dev-only-do-not-override-1adbn6x"
          style="width: 100px;"
        >
          <img
            alt="profile-pic"
            class="ant-image-img rounded-[50%] css-dev-only-do-not-override-1adbn6x"
            src="profile-pic-url"
            width="100"
          />
        </div>
        <div
          class="ant-space css-dev-only-do-not-override-1adbn6x ant-space-vertical ant-space-align-center ant-space-gap-row-small ant-space-gap-col-small mt-2"
        >
          <div
            class="ant-space-item"
          >
            <span
              class="xxl:text-[22px] max-xl:text-[18px] font-extrabold"
            >
              John doe
            </span>
          </div>
          <div
            class="ant-space-item"
          >
            <span
              class="text-base font-bold"
            >
              abc@abc.com
            </span>
          </div>
        </div>
        <div
          class="ant-space css-dev-only-do-not-override-1adbn6x ant-space-horizontal ant-space-align-center ant-space-gap-row-small ant-space-gap-col-small flex items-start text-lg mt-2"
        >
          <div
            class="ant-space-item"
          >
            <table
              class="text-base"
            >
              <tbody>
                <tr>
                  <td
                    class="px-2 py-1 text-gray-500 font-semibold"
                  >
                    Patient Id
                  </td>
                  <td
                    class="px-2 py-1 text-lg"
                  >
                    A01
                  </td>
                </tr>
                <tr>
                  <td
                    class="px-2 py-1 text-gray-500 font-semibold"
                  >
                    Time Slot
                  </td>
                  <td
                    class="px-2 py-1 text-lg"
                  >
                    10:00 am to 11:00 am
                  </td>
                </tr>
                <tr>
                  <td
                    class="px-2 py-1 text-gray-500 font-semibold"
                  >
                    Last Visited
                  </td>
                  <td
                    class="px-2 py-1 text-lg"
                  >
                    28 Dec 2023
                  </td>
                </tr>
                <tr>
                  <td
                    class="px-2 py-1 text-gray-500 font-semibold"
                  >
                    Phone
                  </td>
                  <td
                    class="px-2 py-1 text-lg"
                  >
                    9876543210
                  </td>
                </tr>
                <tr>
                  <td
                    class="px-2 py-1 text-gray-500 font-semibold"
                  >
                    Gender
                  </td>
                  <td
                    class="px-2 py-1 text-lg"
                  >
                    Male
                  </td>
                </tr>
              </tbody>
            </table>
            <div
              class="ant-space css-dev-only-do-not-override-1adbn6x ant-space-horizontal ant-space-align-center ant-space-gap-row-small ant-space-gap-col-small mt-5"
            >
              <div
                class="ant-space-item"
              >
                <button
                  class="ant-btn css-dev-only-do-not-override-1adbn6x ant-btn-default ant-btn-dangerous"
                  type="button"
                >
                  <span>
                    Cancel Appointment
                  </span>
                </button>
              </div>
              <div
                class="ant-space-item"
              >
                <button
                  class="ant-btn css-dev-only-do-not-override-1adbn6x ant-btn-default"
                  type="button"
                >
                  <span>
                    Prescriptions
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
